akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -512,6 +520,20 @@ private[stream] class ActorGraphInterpreter(_initial: GraphInterpreterShell) ext
         false
     }
 
+  //this limits number of messages that can be processed synchronously during one actor receive.
+  private val shortCircuitLimit: Int = _initial.mat.settings.syncProcessingLimit
+  private var shortCircuitBuffer: FixedSizeBuffer[BoundaryEvent] = null
+  private var currentLimit: Int = shortCircuitLimit
+
+  def enqueueToShortCircuit(input: AsyncInput): Boolean = {
+    if (shortCircuitBuffer == null) shortCircuitBuffer = FixedSizeBuffer[BoundaryEvent](shortCircuitLimit)
+    if (!shortCircuitBuffer.isFull && currentLimit > 0) {
```

Comment:
> This won’t work: we cannot execute async callbacks out of order, that would be tremendously confusing for people (and for our own code as well).

Yeah, I actually thought that it is of course possible to reorder since we don't guarantee any ordering, and then realized that in a lot of places we rely on ordering. Fixing that would be actually more costly than simply enforcing it. So let's add a test that checks ordering, and also document this guarantee.